pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -405,52 +405,6 @@ func isSkipped(rprt *ResolvedPipelineRunTask, stateMap map[string]*ResolvedPipel
 	return false
 }
 
-func findReferencedTask(pb string, state []*ResolvedPipelineRunTask) *ResolvedPipelineRunTask {
-	for _, rprtRef := range state {
-		if rprtRef.PipelineTask.Name == pb {
-			return rprtRef
-		}
-	}
-	return nil
-}
-
-// ValidateFrom will look at any `from` clauses in the resolved PipelineRun state
-// and validate it: the `from` must specify an input of the current `Task`. The `PipelineTask`
-// it corresponds to must actually exist in the `Pipeline`. The `PipelineResource` that is bound to the input
-// must be the same `PipelineResource` that was bound to the output of the previous `Task`. If the state is
-// not valid, it will return an error.
-func ValidateFrom(state PipelineRunState) error {
```

Comment:
This is the `PipelineSpec` `validateFrom` function.

```
func validateFrom(tasks []PipelineTask) error {
	taskOutputs := map[string][]PipelineTaskOutputResource{}
	for _, task := range tasks {
		var to []PipelineTaskOutputResource
		if task.Resources != nil {
			to = make([]PipelineTaskOutputResource, len(task.Resources.Outputs))
			copy(to, task.Resources.Outputs)
		}
		taskOutputs[task.Name] = to
	}
	for _, t := range tasks {
		if t.Resources != nil {
			for _, rd := range t.Resources.Inputs {
				for _, pb := range rd.From {
					outputs, found := taskOutputs[pb]
					if !found {
						return xerrors.Errorf("expected resource %s to be from task %s, but task %s doesn't exist", rd.Resource, pb, pb)
					}
					if !isOutput(outputs, rd.Resource) {
						return xerrors.Errorf("the resource %s from %s must be an output but is an input", rd.Resource, pb)
					}
				}
			}
		}
	}
	return nil
}
```
In the `PipelineSpec` validation, we don't check the `sameBinding` case :thinking: I am not entirely sure what this check was about but we may be missing something…

Reply:
Looking at the test cases for the sameBinding case i.e. the ones that test that the error contains the word ambiguous, there are 3 test cases:

1. from tries to reference input  - covered by spec.Validate
2. "from resource doesn't exist" - covered by spec.Validate	
3. "from is bound to different resource" - need to double check what this  case is doing 🤔

Reply:
Ok, so case 3 is from say task B has a from that says use resource "foo" from Task A while Task A produces a resource named "bar". I think pipelineSpec.Validate [covers this](https://github.com/tektoncd/pipeline/blob/master/pkg/apis/pipeline/v1alpha1/pipeline_validation.go#L71). So, I think all 3 cases of "same binding" are covered